subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.plan;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import org.gradle.internal.snapshot.CaseSensitivity;
+import org.gradle.internal.snapshot.ChildMap;
+import org.gradle.internal.snapshot.ChildMapFactory;
+import org.gradle.internal.snapshot.EmptyChildMap;
+import org.gradle.internal.snapshot.VfsRelativePath;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+
+public class RelatedLocations {
+    private volatile RelatedLocation root = new DefaultRelatedLocation(EmptyChildMap.getInstance(), ImmutableList.of());
```

Comment:
I wonder if it's better like this, or if we should be using `DefaultRelatedLocation` as the thing. For the VFS I see a reason for keeping a mutable wrapper around the immutable data structure, because we need to access the mutable thing from different places. Here we only have one place where we update this thing, and I think we could simplify significantly if we dropped the mutable wrapper.

Reply:
We still have multiple places where we update the thing: in the execution plan and from the task executor. So I think we need a mutable wrapper to keep the right reference in two places.